src/JSInterop/Microsoft.JSInterop/src/Infrastructure/ByteArrayJsonConverter.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using System.Threading;
+
+namespace Microsoft.JSInterop.Infrastructure
+{
+    internal sealed class ByteArrayJsonConverter : JsonConverter<byte[]>
+    {
+        // Atomically incrementing unique identifier for serializing byte arrays.
+        private static long _byteArrayId;
+
+        internal static readonly JsonEncodedText ByteArrayRefKey = JsonEncodedText.Encode("__byte[]");
+
+        public ByteArrayJsonConverter(JSRuntime jSRuntime)
+        {
+            JSRuntime = jSRuntime;
+        }
+
+        public JSRuntime JSRuntime { get; }
+
+        public override bool CanConvert(Type typeToConvert) => typeToConvert == typeof(byte[]);
+
+        public override byte[]? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (JSRuntime.ByteArraysToBeRevived.Count == 0)
+            {
+                throw new JsonException("ByteArraysToBeRevived is empty.");
+            }
+
+            long? byteArrayIndex = null;
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndObject)
+            {
+                if (reader.TokenType == JsonTokenType.PropertyName)
+                {
+                    if (byteArrayIndex is null && reader.ValueTextEquals(ByteArrayRefKey.EncodedUtf8Bytes))
+                    {
+                        reader.Read();
+                        byteArrayIndex = reader.GetInt64();
+                    }
+                    else
+                    {
+                        throw new JsonException($"Unexpected JSON property {reader.GetString()}.");
+                    }
+                }
+                else
+                {
+                    throw new JsonException($"Unexpected JSON Token {reader.TokenType}.");
+                }
+            }
```

Comment:
There's a simpler way to do this, which involves defining a private type
```csharp
public struct ByteArrayRef
{
  [JsonPropertyName("__byte[]")]
  public int Id { get; set; }
}
```

And calling `JsonSerializer.Deserialize<ArrayRef>(reader, options)` which saves you from dealing with the low level stuff

Reply:
Interesting, cool idea